src/operation.rs

```diff
@@ -71,9 +71,35 @@ const SERVER_4_2_0_WIRE_VERSION: i32 = 8;
 const SERVER_4_4_0_WIRE_VERSION: i32 = 9;
 // The maximum number of bytes that may be included in a write payload when auto-encryption is
 // enabled.
-const MAX_ENCRYPTED_WRITE_SIZE: u64 = 2_097_152;
+const MAX_ENCRYPTED_WRITE_SIZE: usize = 2_097_152;
 // The amount of overhead bytes to account for when building a document sequence.
-const COMMAND_OVERHEAD_SIZE: u64 = 16_000;
+const COMMAND_OVERHEAD_SIZE: usize = 16_000;
+
+pub(crate) enum OperationResponse<'a, O> {
+    Sync(Result<O>),
+    Async(BoxFuture<'a, Result<O>>),
+}
+
+impl<'a, O> OperationResponse<'a, O> {
+    /// Returns the sync result contained within this `OperationResponse`. Use responsibly, when it
+    /// is known that the response is not async.
+    fn get_sync_result(self) -> Result<O> {
+        match self {
+            Self::Sync(result) => result,
+            Self::Async(_) => Err(Error::internal(
+                "get_sync_result was called on an async response",
+            )),
+        }
+    }
+}
+
+macro_rules! handle_response_sync {
```

Comment:
This macro allows using `?` throughout the impl block, which won't work in something like `OperationResponse::Sync({...})` because the return type of `handle_response` is no longer a result/option. I was having trouble making this work in a macro without the closure indirection, although there might be a simpler way. I did just mess around with this a little more and the following will also work in the body of `handle_response`:

```
let result: Result<_> = (|| {
   // parse and return result
})();
OperationResponse::Sync(result)
```

Neither option is that great. I lean slightly more towards the macro as the create/immediately call closure syntax without it is a little funky. (This will also go away if we decide to add `Send + Sync` to `T`.)

Reply:
Aha, use of `?` is what I was missing, sticking with the macro sounds fine to me.  Once [try blocks](https://github.com/rust-lang/rust/issues/31436) are usable that will solve this, but who knows how long that'll be.